### PR TITLE
Adding signup snippet.

### PR DESCRIPTION
**Pull request by ScottDowne**

### CODE REVIEW

**Comment by ScottDowne** [comment]
@glogiotatidis Most of the video and email code was adapted from other snippets.

The bulk of the UI was adapted from EOY snippets.

I'll also be around pretty early tomorrow, to respond to any questions you have.

File: campaigns/signup-snippet/snippet.html [hunk 1]
```diff
@@ -0,0 +1,1458 @@
+<div class="snippet" id="signup-snippet-{{ snippet_id }}">
+  <style type="text/css">
+    @font-face {
+      font-family: 'fontello-{{ snippet_id }}';
```
Comment (glogiotatidis) [inline campaigns/signup-snippet/snippet.html]: TIP:  Uniquely identifying elements using `{{ snippet_id }} ` is not required anymore.

File: campaigns/signup-snippet/snippet.html [hunk 2]
```diff
@@ -0,0 +1,1458 @@
...
+    </div>
+    <div class="success-message hidden">
+      <p>{{ successMessage }}</p>
+    </div>
+    <div class="form-container">
+      <div class="error-message basket-error hidden">
+        <div class="error-container">{{ basketError }}</div>
+      </div>
+      <div class="row">
+        <input form="form" class="email-input" type="email" required placeholder="{{ emailPlaceholder }}"></input>
```
Comment (glogiotatidis) [inline campaigns/signup-snippet/snippet.html]: Invalid XML. Set `required="required"`

File: campaigns/signup-snippet/snippet.html [hunk 3]
```diff
@@ -0,0 +1,1458 @@
...
+      </div>
+      <div class="row">
+        <input class="first-name-input" type="text" placeholder="{{ firstnamePlaceholder }}"></input>
+      </div>
+      <div class="row">
+        <select class="country-input" required>
```
Comment (glogiotatidis) [inline campaigns/signup-snippet/snippet.html]: Invalid XML.

File: campaigns/signup-snippet/snippet.html [hunk 4]
```diff
@@ -0,0 +1,1458 @@
...
+            Zambia
+          </option>
+
+          <option value="ZW">
+            Zimbabwe
+          </option>
+        </select>
+      </div>
+      <div class="row">
+        <label><input form="form" class="privacy-checkbox" required type="checkbox"></input>{{ privacyPolicy|safe }}</label>
```
Comment (glogiotatidis) [inline campaigns/signup-snippet/snippet.html]: Invalid XML

File: campaigns/signup-snippet/snippet.html [hunk 5]
```diff
@@ -0,0 +1,1458 @@
...
+      </div>
+      <div class="row">
+        <input form="form" class="email-input" type="email" required placeholder="{{ emailPlaceholder }}"></input>
+      </div>
+      <div class="row">
+        <input class="first-name-input" type="text" placeholder="{{ firstnamePlaceholder }}"></input>
+      </div>
+      <div class="row">
+        <select class="country-input" required>
+          <option value="" selected>
```
Comment (glogiotatidis) [inline campaigns/signup-snippet/snippet.html]: Invalid XML

**Comment by glogiotatidis** [comment]
As a general rule going forward please provide a list of the variables in the template, along with their type and description text.

File: campaigns/signup-snippet/snippet.html (hunk 1 above)
Comment (ScottDowne) [inline campaigns/signup-snippet/snippet.html]: Good to know!

**Comment by ScottDowne** [comment]
I tested with this set of variables:

        "text": "Mozilla stands up for online privacy. Will you join us? Sign up to get important email updates and opportunities to stand with us.",
        "emailPlaceholder": "Email (required)",
        "firstnamePlaceholder": "First Name",
        "countryPlaceholder": "Country",
        "lang": "en-US",
        "privacyPolicy": "I’m okay with Mozilla handling my info as explained in <a href=\"https://www.mozilla.org/en-US/privacy/websites/\">this Privacy Notice</a>.",
        "submit": "Submit",
        "blockable": true,
        "basketHost": "https://basket-dev.allizom.org",
        "closeButtonTitle": "close",
        "basketError": "There was an unexpected error, please try again later.",
        "successMessage": "Great! Look for a confirmation email in your inbox to complete your subscription to our email list.", 
        "newsletter": "mozilla-foundation",
        "video_url": "http://video.webmfiles.org/big-buck-bunny_trailer.webm",
        "video_poster": "http://jplayer.org/video/poster/Big_Buck_Bunny_Trailer_480x270.png"

I'm not super sure how I should add type and description text to them.

**Comment by glogiotatidis** [comment]
> I'm not super sure how I should add type and description text to them.

I'm researching ideas on how we can do this part of the snippet and have the backend automatically load and configure the variables. 

Meanwhile you can add these as comments here.

**Comment by glogiotatidis** [comment]
screenshot from 2016-02-04 16-39-23

Some text and elements are off when we bring this into about:home. About:home is a very special page. I know that developing a snippet as a regular page is waaay easier but please don't forget to actually test it in real conditions.

**Comment by glogiotatidis** [comment]
Also you need to apply a fix similar to this to prevent search box from taking focus in current fx versions.
https://github.com/mozilla/snippets/commit/9f3d68bda0eb2a9c1c023cab6dcaff3b4ef6fbd7

**Comment by glogiotatidis** [comment]
Are you using any features what are available to recent Fx versions only?

**Comment by glogiotatidis** [comment]
OK this is for a first pass. The snippet seems to work apart from the issues reported. I'll be available for another 4 hours today, feel free to ping me when you update!

**Comment by ScottDowne** [comment]
> Are you using any features what are available to recent Fx versions only?

No, nothing new.

I tested it as low as FireFox 13 out of curiosity. The error dialog looked vastly different and some of the placeholder alignment was off, but no broken functionality. 12 seemed pretty busted. Is there a range of browser versions the snippet supports? I figure the alignment in 13 is totally fixable.

ff13
ff13-2

**Comment by glogiotatidis** [comment]
As a general rule Fx 29+ is good enough. I try to have some basic snippets supporting Fx 15+. So you're good! :+1:

**Comment by ScottDowne** [comment]
@glogiotatidis If I forgot my admin for my local django is there an easy way to reset it without setting up the whole stack again?

**Comment by ScottDowne** [comment]
Hm, next issue.

I'm trying to set updateUrl in about:config to http://127.0.0.1:8000/

When I then close the browser and refresh snippets with window.gSnippetsMap.clear();

I get the error: Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at http://127.0.0.1:8000/. (Reason: CORS header 'Access-Control-Allow-Origin' missing).

I fixed this during end of year, but now hitting it again and cannot seem to hit the same magic to fix it.

I remember it not being an obvious CORS thing, I remember it being something else.

Is this a common gotchya in setting a new snippet source?

**Comment by ScottDowne** [comment]
Ha, got it!

I don't know what I did...

**Comment by ScottDowne** [comment]
The weird styling when in about:home has something to do with some of my styles being ignored, or otherwise replaced with the @font-face rules. Not sure yet why, debugging.

**Comment by ScottDowne** [comment]
I think I got rid of the CSS issues.

**Comment by ScottDowne** [comment]
Updates, still need to figure out the removal of the id, but I left that for the end in hopes I would have it done in less than four hours, but I think it might be too late anyway :)

**Comment by glogiotatidis** [comment]
> Updates, still need to figure out the removal of the id, but I left that for the end in hopes I would have it done in less than four hours, but I think it might be too late anyway :)

Doesn't hurt to keep snippet_id if that's what you mean. In the future you just don't need to add it

**Comment by ScottDowne** [comment]
>>    Updates, still need to figure out the removal of the id, but I left that for the end in hopes I would have it done in less than four hours, but I think it might be too late anyway :)

> Doesn't hurt to keep snippet_id if that's what you mean. In the future you just don't need to add it

K, let's leave it as is. Removing it makes me a bit nervous this close to being finished.

**Comment by glogiotatidis** [comment]
Nits:
 - Success message is not aligned in center
 - I see that we load a 50kb font to display a spinner. This shouldn't be the case with snippets as there are reasons to keep them as small as possible. Since this is strict deadline if you don't find another way to implement this it's ok but going forward we should find a way.

**Comment by glogiotatidis** [comment]
From my end this looks ok. If you want to fix the last two nits before you go, otherwise i'll upload as is. 

Thanks @ScottDowne :cake:

**Comment by ScottDowne** [comment]
I initially got the font from here: https://github.com/mozilla/snippets/blob/master/campaigns/10th-anniversary-video/10th-anniversary-video.html#L4

It contained just a play and pause button icon, I then added the spinner. From my calculations it went from 2.7 KB to 3.1 KB.

I totally see no harm in it being smaller, so I'll see what I can do.

**Comment by glogiotatidis** [comment]
> It contained just a play and pause button icon, I then added the spinner. From my calculations it went from 2.7 KB to 3.1 KB.

Only 3kb? I must have measured the wrong thing then. It's ok, thnx!

**Comment by ScottDowne** [comment]
3.1 KB not 3.1 kb

I think 3.1 KB is 24.8 kb?

**Comment by ScottDowne** [comment]
I can easily make a css only spinner, something like one of these: http://projects.lukehaas.me/css-loaders/

**Comment by glogiotatidis** [comment]
> I can easily make a css only spinner, something like one of these: http://projects.lukehaas.me/css-loaders/

I find this a better approach if you have the time.

**Comment by ScottDowne** [comment]
Yup sounds good. Should only be a few mins. I'll still be using an icon for
replay but might make a CSS play icon too.

**Comment by ScottDowne** [comment]
Been trying to reproduce the success message not centered, it's centered for me.

**Comment by ScottDowne** [comment]
I think I figured out the centering bit. For some reason I thought you meant horizontal centering, but I noticed the vertical centering is off, so it's a really good bet that's what you're seeing. Easy to fix!

**Comment by ScottDowne** [comment]
@glogiotatidis I have:
- updated the spinner to be css
- fixed the alignment of the success message
- fixed a bug where country wasn't doing stopPropagation
- and I also found a bug where the submit button wasn't being turned off.

**Comment by glogiotatidis** [comment]
Nice work @ScottDowne . Thanks for your patience :) :cake:

**Pull request closed without merging**